Network latency improvements

jukkar opened the pull request:
This PR is divided to two parts but as they are related / depend each other, I am sending all commits in the same PR.

Main change here is to avoid creating TX thread which saves RAM and improves message throughput when sending data. We can only do this if userspace support is disabled. In my tests this improved UDP TX sending time by 30%, and if RX thread is also disabled, RX receive time by 14% in qemu_x86 and echo-server/client. Removing RX thread is not recommended because then most of the RX data path is run in ISR context.

Also add a new sample application for testing SO_TXTIME that can be used to send packets in certain time in the future (requires support from Ethernet driver). This can be used in TSN testing.

jukkar committed:
drivers: eth: e1000: Add simulated PTP clock device

For testing purposes, add simulated PTP clock device to e1000
Ethernet driver that is used in qemu_x86 board. The PTP clock
does nothing useful as there is no real hw behind this device.
We just emulate the clock in order to do some SO_TXTIME testing.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
samples: net: socket: txtime: Simple app for SO_TXTIME

This application shows how to use SO_TXTIME option, which can be
used to specify the time when the network packet should be sent
out by the network device driver.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: Push highest priority net_pkt directly to driver

If user has set the priority of the sent net_pkt to highest
priority (NET_PRIORITY_CA) and enabled CONFIG_NET_TC_SKIP_FOR_HIGH_PRIO
option, then push that packet directly to driver instead of TX queue.
This will make the TX sending latency smaller for the high priority
packet. This is not enabled by default.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
drivers: eth: gmac: Separate PTP clock from gPTP support

It is possible that user wants to access PTP clock but does
not need gPTP support. The networking txtime sample does exactly
this.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: pkt: Set the pkt creation time more accurately

Set the net_pkt creation time just before the actual net_pkt
is allocated in order to get more accurate information for
statistics.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: Enable running without TX or RX threads

Set the default behaviour of the networking subsystem so that
no TX or RX threads are created. This will save RAM as there
is no need to allocate stack space for the RX/TX threads.
Also this will give small improvement to network packet latency
shown here:
* with 1 traffic class (1 TX and RX thread)

Avg TX net_pkt (42707) time 60 us	[0->22->15->22=59 us]
Avg RX net_pkt (42697) time 36 us	[0->10->3->12->7=32 us]

* with 0 traffic classes (no TX and RX threads)

Avg TX net_pkt (41608) time 42 us	[0->21->20=41 us]
Avg RX net_pkt (41593) time 31 us	[0->9->12->8=29 us]

In this qemu_x86 test run, 40k UDP packets was transferred between
echo-server and echo-client. In TX the speed increase was 30% and
in RX it was 14%.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: If userspace support is enabled, then we need TX/RX threads

If user enables CONFIG_USERSPACE, then at least one TX or RX thread
is needed to isolate the application from the kernel space components.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
drivers: eth: mcux: Separate PTP clock from gPTP support

It is possible that user wants to access PTP clock but does
not need gPTP support. The networking txtime sample does exactly
this.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
tests: net: ptp: Compile only on relevant boards

The tests need gPTP support so set the options and boards
properly.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: pkt: Have separate create time for net_pkt

This value is used to measure the RX/TX statistics. The previous
use of the timestamp field did not work in RX path as the timestamp
value could be overwritten by the driver if gPTP timestamping
is enabled. So to fix the RX statistics, use a separate field
for the create time.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
tests: net: poll: Increase ztest thread stack size

The socket poll test was running out of ztest stack so increase
it by 256 bytes.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

rlubos commented on samples/net/sockets/txtime/src/main.c:
```diff
@@ -0,0 +1,536 @@
+
+		ret = sendmsg(data->sock, &msg, 0);
+		if (ret < 0) {
+			if (errno != ENOMEM) {
```
Why is ENOMEM expected here?

jukkar replied:
> Why is ENOMEM expected here?

The application does not have much buffers configured so it might run out of them, if sending time is longer in the future and if the sending period is very small. So for this application, it is not a big issue if we run out of buffers, we can ignore this and retry later.

rlubos commented on samples/net/sockets/txtime/src/main.c:
```diff
@@ -0,0 +1,536 @@
+			}
+		}
+
+		txtime += delay + interval;
```
The delay will accumulate over time (actually twice as fast given the `delay` is also added to the `tx_time` in line 191). I think this can also lead to the TX packet starvation, given that we only sleep for the `interval` time, yet the packet is sent every `interval + 2 * delay`. 

My thinking is that we should rather increment the `txtime` only with the value of the `interval` (`txtime += interval;`) and add a delay at line 192 (`*(uint64_t *)CMSG_DATA(cmsg) = txtime + delay;`) so that the delay does not accumulate over time, and we send a packet every `interval` msecs (with the delay).

jukkar replied:
> I think this can also lead to the TX packet starvation

Indeed, actually that can explain the ENOMEM I had to add.

rlubos reviewed (commented):
Nice improvements, thanks. I think that the possibility to drop the RX/TX threads would be useful for OpenThread, where we have a separate thread to process incoming/outgoing packets anyway, so we could save some memory here.

jukkar commented on include/net/net_pkt.h:
```diff
@@ -104,6 +104,9 @@ struct net_pkt {
 		/** Timestamp if available. */
 		struct net_ptp_time timestamp;
```
> Aren't the `timestamp` and `create_time` separate features now?

Good point, I will refactor the code as suggested.

jukkar committed:
net: pkt: Separate the timestamp and TX/RX stats configs

The timestamp is no longer depending on TX/RX time config
options so move it to separate settings.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar commented:
Updated according to comments and rebased to latest upstream.

rlubos reviewed (approved):
LGTM

jukkar commented:
Fixed merge conflict.

It would be nice to get this to 2.6, so any second review would be much appreciated.

pfalcon commented on subsys/net/ip/Kconfig:
```diff
@@ -140,9 +140,9 @@ config NET_SHELL_DYN_CMD_COMPLETION
-	default 1
-	range 1 NET_TC_NUM_PRIORITIES if NET_TC_NUM_PRIORITIES<=8
-	range 1 8
+	default 0
```
Are we really ready to change the default like that? How will we test config with threads?

pfalcon commented:
The changes look ok, but IMHO too much packed into one PR, it could have split into 2-3 PRs to ease review/merging/testing.

My concern is switching default to no RX/TX threads, given following:
> "Disabling RX thread means that the
	  network device driver, that is typically running in IRQ context, will
	  handle the packet all the way to the application. This might cause
	  other incoming packets to be lost if the RX processing takes long
	  time."

Besides basic question "do existing samples work ok with that?", there's also "how do we test config with threads enabled?". I see that they're enabled with CONFIG_USERSPACE, but I'm not sure is we use that for networking tests either.

All in all, I'd prefer to give +1 only after testing my default test apps (dump_http_server/big_http_dl).

(I'm writing this with understanding that there're 2 votes already, and it's going to be merged. That's fine, again, I don't see anything wrong there.)

jukkar commented:
> Besides basic question "do existing samples work ok with that?", there's also "how do we test config with threads enabled?". I see that they're enabled with CONFIG_USERSPACE, but I'm not sure is we use that for networking tests either.

The only major change is that we avoid creating TX thread. What happens in practice, is that when sending data, instead of putting net_pkt to a queue, the net_pkt is passed to the L2 directly. I do not encounter any problems with that.
In RX side, nothing is changed by default, but if user wants, it is possible to avoid creating RX thread.

What does "how do we test config with threads enabled" mean, I do not understand your question.

pfalcon commented:
>  In RX side, nothing is changed by default, but if user wants, it is possible to avoid creating RX thread.

Aha, I missed that default change to 0 only for NET_TC_TX_COUNT, NET_TC_RX_COUNT stays at 1.

> What does "how do we test config with threads enabled" mean, I do not understand your question.

I mean, we know the behavior of the stack (and apps using it) with RX/TX threads enabled, and it would be nice to get more people testing it on various HW before switching over. I agree that changing just NET_TC_TX_COUNT seems low-risk enough, so let me +1 this now. I'll test as time permits, or otherwise as part of the usual release testing.

vandita5 commented:
@jukkar will this change be included in 2.6.0?

jukkar commented:
Yes, it is part of the coming 2.6 release.
